Some (sadly breaking) improvements to the Fluid API

gigaherz opened the pull request:
* Replaced the string id on the FluidAttributes with a reference to the owner fluid.
* Added a boolean to distinguish when fluids want to use biome-based water coloring.
* Added logic to automatically gather fluid textures into the list of textures to bake.

gigaherz committed:
Some (sadly breaking) improvements to the API, after feedback and concerns provided by users:
* Replaced the string id on the FluidAttributes with a reference to the owner fluid.
* Added a boolean to distinguish when fluids want to use biome-based water coloring.
* Added logic to automatically gather fluid textures into the list of textures to bake.

pupnewfster commented:
I am not positive, but I think water and lava *should* override the translation key as I don't believe there are any lang keys in vanilla for fluids that start with `fluid`. I do agree that the way this PR is introducing for the default translation key is better though.

Edit: I am blind already is overridden.

Ommina commented:
Is the expectation here that the mod author will create a new class extending FlowingFluid for each and every fluid they create?  It feels like any mod that adds more than a couple fluids are going to rapidly get lost in a sea of copy/paste code.

I've been working on generalize FlowingFluid, but it isn't pretty, and the change to FluidAttributes taking a Fluid reference in this PR is going to make it even less so.

gigaherz commented:
Ultimately, yes, like we do for blocks and items. I'm working toward making it possible to create a more flexible system using suppliers, but my focus is on the base work in building up the vanilla system.

I was considering making something equivalent to Item.Properties and Block.Properties, for constructing a FlowingFluid (would have to be a new ForgeFlowingFluid class) so that implementing the new fluids doesn't require spelling out all the boilerplate, but that will come later.

gigaherz committed:
Patched BucketItem and FlowingFluidBlock to avoid eager access to the fluid objects.

Yamahari commented on patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch:
```diff
@@ -1,6 +1,71 @@
+    @OnlyIn(Dist.CLIENT)
+    public boolean func_200122_a(BlockState p_200122_1_, BlockState p_200122_2_, Direction p_200122_3_) {
+-      return p_200122_2_.func_204520_s().func_206886_c().func_207187_a(this.field_204517_c) ? true : super.func_200124_e(p_200122_1_);
++      return p_200122_2_.func_204520_s().func_206886_c().func_207187_a(getFluid()) ? true : super.func_200124_e(p_200122_1_);
```
while you're at it, this could be simplified to 
`p_200122_2_.func_204520_s().func_206886_c().func_207187_a(getFluid()) || super.func_200124_e(p_200122_1_);`

gigaherz replied:
It could, but I think that would also make the patch less obvious, I'd rather leave it as it is, so that people see what REALLY changed in that patch.

mezz commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -157,11 +157,7 @@ protected FluidAttributes(Builder builder)
-
-    public final String getName()
-    {
-        return this.fluidName;
```
to avoid a breaking change here, can this become
```java
return owner.getRegistryName().toString();
```
and deprecate it?

KingLemming replied:
Do we actually have an RB at this point? I think we're early enough in dev that we shouldn't have to overly worry about deprecation. I'd rather it break and people use it correctly.

gigaherz replied:
That is my thinking. On top of that, this PR would have breaking changes regardless, so the value of this deprecated method that I just added last week seems too low to be worth keeping it.

I was hoping to get this merged quickly to avoid a lot of people having to deal with re-releasing updates, but I really want Lex's input on this.

KingLemming replied:
Yeah, if changes are *already* broken then I say maximize all possible breakage now. Let's make 1.14 as clean as possible. No deprecations, etc. Damn the torpedoes.

mezz replied:
Ok, makes sense to me

LexManos replied:
We are holding off a RB explicitly for this fluid system, so that it can do everything correctly and THEN we can stabilize the binary version. So once this is done, we'll worry about compat.

mezz commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -291,6 +287,11 @@ public int getColor()
         return color;
     }
 
+    public boolean getUseBiomeWaterColor()
```
This flag is too specific to be here, it is not at the right level of abstraction.

Instead, I think we should create subclass of `FluidAttributes` just for water that overrides `FluidAttributes#getColor(IEnviromentBlockReader, BlockPos)` to call `BiomeColors.getWaterColor`.
That should simplify the patch a little as well.

gigaherz replied:
I was initially thinking I don't like the idea, but actually, after poking at it in the code, I do.

It actually doesn't change the patch that much, but I guess the logic in FluidBlockRenderer is simpler now.

gigaherz committed:
Move the biome color to a dedicated FluidAttributes.Water subclass.

Alpvax commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -458,4 +444,36 @@ public FluidAttributes build()
+            @Override
+            public FluidAttributes build()
+            {
+                return new Water(this);
```
Could we not add a simpler way of subclassing the attributes without subclassing the builder. e.g. by optionally passing in a factory to `build` (i.e. the water builder would become `build(Water::new)`) or to the builder constructor, or adding a new builder function `setAttributesClass` or something.
That would make it easier for other mods which wanted to add a custom colour map, or otherwise override one of the FluidAttributes implementations.

LexManos replied:
This should be a simple enough change to add, you should do it.

gigaherz committed:
Add burn time value to fluids, loosely based on the proposal from #4678.

gigaherz committed:
Rename back to fuel.

ichttt commented:
is this done? Would be a good time to merge now, as we just had another breaking change

gigaherz commented:
It is. We haven't had any other fluids issue. I just wanted Lex's confirmation that this is good to merge, since it adds quite a few patches to 2 classes.

LexManos commented:
I was waiting on you to tell me things are ready to look at.

LexManos commented on patches/minecraft/net/minecraft/fluid/Fluid.java.patch:
```diff
@@ -24,29 +24,34 @@
 +    */
-+   protected net.minecraftforge.fluids.FluidAttributes createAttributes(Fluid fluid)
++   protected net.minecraftforge.fluids.FluidAttributes createAttributes()
 +   {
```
This would be better off moving to a ForgeHooks method, just due to the shear size, and uglyness that is due to being in a patch.

gigaherz replied:
Done.

LexManos commented on patches/minecraft/net/minecraft/block/FlowingFluidBlock.java.patch:
```diff
@@ -1,6 +1,71 @@
+ 
+    public int func_149738_a(IWorldReader p_149738_1_) {
+-      return this.field_204517_c.func_205569_a(p_149738_1_);
++      return getFluid().func_205569_a(p_149738_1_);
```
Use our field to method trick we use for ItemStack/PotionEffect for this. Would vastly simplify the patch set and maintainability.

gigaherz replied:
Done (tentative).

LexManos commented on patches/minecraft/net/minecraft/item/BucketItem.java.patch:
```diff
@@ -1,33 +1,104 @@
 -      SoundEvent soundevent = this.field_77876_a.func_207185_a(FluidTags.field_206960_b) ? SoundEvents.field_187627_L : SoundEvents.field_187624_K;
-+      SoundEvent soundevent = field_77876_a.getAttributes().getEmptySound();
-+      if(soundevent == null) soundevent = this.field_77876_a.func_207185_a(FluidTags.field_206960_b) ? SoundEvents.field_187627_L : SoundEvents.field_187624_K;
++      SoundEvent soundevent = getFluid().getAttributes().getEmptySound();
```
Perhaps would be worth doing the same runtime redirect here as well to save a lot of patches

LexManos commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -141,9 +132,20 @@
+     * For reference, smelting one item takes 200 ticks, one piece of coal = 1600 ticks,
+     * and a bucket of lava lasts for 20000 ticks.
+     */
+    protected int fuelTime = 0;
```
javadocs are helpful on PUBLIC methods, so add this information to the getter as well.

gigaherz replied:
Removed the fuelTime stuff instead.

LexManos commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -338,22 +353,31 @@ public SoundEvent getEmptySound()
     public static class Builder
     {
-        private final String name;
+        private final Fluid owner;
```
Why does this need to be the fluid instance? I can see this being a chicken-egg style thing. 
Have you tried using a registry delegate or something that isn't the Fluid directly.
A delegate would silently support registry replacements. But doesn't solve the chicken/egg thing off hand.
All you seem to use it for is the name, what was the logic behind this change?

gigaherz replied:
My thinking was that it may be useful for the people subclassing FluidAttributes, but on second thought, all methods that are worth overriding either have a FluidStack or a FluidState, meaning both have a .getFluid() method available, so the field could be simplified to a ResourceLocation.

gigaherz replied:
Removed the pointless field. The builder function still takes a Fluid, for convenience, but it's only used to get the registry name from.

gigaherz committed:
Address Lex's comments

gigaherz committed:
Partially revert BucketItem and FlowingFluidBlock patches in favor of a transformer.

tterrag1098 commented:
I would like to have a bit more discussion around the "fuel time" feature before we move forward. 

The main problem I see with this feature is that "fuel time" is not an intrinsic property of a fluid, unlike temperature or viscosity. How a given generator creates power from a liquid is totally open ended. So I propose retooling this feature as a more intrinsic property with a better defined meaning.

It could instead be defined as "energy density" i.e. the max amount of energy stored in the fluid, perhaps represented in FE/mB. This would define only the *maximum* amount of energy that can be generated by a fluid, but not *how* that energy is generated. For instance both lava and gasoline could implement this data point, but it's up to the generator to decide if they can use the specific fluid, perhaps by using tags.

LexManos commented:
Then it's just a naming change. As it was intended that 'fuelTime' was only meant to be a seed value, it was up to other mechanics to determine if it could be burnt, or exploded, or chilled, or whatever mechanic you use to generate the energy.
The problem with extending things to FE/mB is that it then adds a constant value to furnace. And that's a odd conversion.
But as a community we already have the standard and use case of fluids as a furnace fuel. 
So I would still represent it as the 2000 lava has, the 'burntime' and let the consumers define their ratios for length/FE/whatever.

gigaherz committed:
Get rid of the unused owner field.

gigaherz committed:
New batch of improvements:
- We decided it's best NOT to have the fuelTime value at all, since it's too unclear and people are saying they can't trust it.
- Tentative version of FlowingFluidBlock and BucketItem alternative constructors that take a Supplier<Fluid>, it is intended to be used with a RegistryObject<Fluid>.
- Addressed some typos.

gigaherz commented:
At this point, this should be working, and could get a new pass of reviews.
The FlowingFluidBlock and BucketItem taking a supplier is done with the assumption that certain changes to `RegistryObject<T>` make it into the code, if not, those constructors may need to be reconsidered.

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -831,6 +831,30 @@ public static LootTable loadLootTable(Gson gson, ResourceLocation name, JsonObje
+                    new net.minecraft.util.ResourceLocation("white"),
+                    new net.minecraft.util.ResourceLocation("white"))
+                    .translationKey("block.minecraft.air")
+                    .color(0).density(0).temperature(0).luminosity(0).viscosity(0).density(0).build();
```
Densty is set twice in this line

gigaherz replied:
oops!

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -363,12 +353,13 @@ public static Builder builder(String name, ResourceLocation stillTexture, Resour
 
-        protected Builder(String name, ResourceLocation stillTexture, ResourceLocation flowingTexture) {
-            this.name = name.toLowerCase(Locale.ENGLISH);
+        protected Builder(Fluid owner, ResourceLocation stillTexture, ResourceLocation flowingTexture, Function<Builder,FluidAttributes> factory) {
```
As owner is only used for getting the registry name for the translation key, would it make sense to have two constructors, one that takes Fluid and one that takes a ResourceLocation (being the registry name). I don't fully remember the scope of how FluidAttributes gets made/used so this may not even make sense to do. (Or maybe it makes sense to change the protected builder to *only* ResourceLocation for registry name and make the `public static Builder builder(` above pass the registry name to this constructor.

gigaherz replied:
`FluidAttributes` construction is something that should only happen within the `createAttributes` method of the Fluid class, so in principle 100% of the times the given value in that parameter will be the `this`, within `createAttributes`. The main reason the builder isn't given privately to the createAttributes method directly, is that mod fluids might want to construct `FluidAttributes` instances with extra parameters not represented in the default attributes.

gigaherz committed:
Typo

alexbegt commented:
This isn't that important but should the comment in the fieldtomethodtransformers.js say 'potion' for the Fluid block and Bucket item?

gigaherz commented:
No, no it shouldn't. :P

gigaherz committed:
Another nitpick

ichttt commented on src/main/resources/META-INF/fieldtomethodtransformers.js:
```diff
@@ -19,7 +19,7 @@ function initializeCoreMod() {
                 var asmapi=Java.type('net.minecraftforge.coremod.api.ASMAPI')
-                var fn = asmapi.mapField('field_204517_c') // potion field - remap to mcp if necessary
+                var fn = asmapi.mapField('field_204517_c') // fluid field - remap to mcp if necessary
                 asmapi.redirectFieldToMethod(classNode, fn, asmapi.mapMethod('getFluid'))
```
if getFluid is a method that has no srg name (i.e. added by a forge patch),  you can just call `asmapi.redirectFieldToMethod(classNode, fn, 'getFluid')`, as it does not need to be remapped

gigaherz replied:
Oops, completely missed that lol.
/me hides in shame
Will fix when I'm back home later

ichttt commented on src/main/resources/META-INF/fieldtomethodtransformers.js:
```diff
@@ -31,7 +31,7 @@ function initializeCoreMod() {
                 var asmapi=Java.type('net.minecraftforge.coremod.api.ASMAPI')
-                var fn = asmapi.mapField('field_77876_a') // potion field - remap to mcp if necessary
+                var fn = asmapi.mapField('field_77876_a') // containerFluid (wrongly named containedBlock) field - remap to mcp if necessary
                 asmapi.redirectFieldToMethod(classNode, fn, asmapi.mapMethod('getFluid'))
```
same here

ichttt reviewed (changes requested):
Sorry, some nitpicks from me regarding the asm as well

gigaherz committed:
More oppsies fixed

gigaherz committed:
Implement ForgeFlowingFluid, as a more user-friendly way to construct a new fluid.

gigaherz commented:
I have added what I think is a really handy helper class for implementing fluids. I'm now going to sleep, but I plan on addressing any concern when I wake up, and then merge this PR, so if you have any comment abut the ForgeFlowingFluid, now's the time.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java:
```diff
@@ -0,0 +1,243 @@
+    @Override
+    protected boolean func_215665_a(IFluidState p_215665_1_, IBlockReader p_215665_2_, BlockPos p_215665_3_, Fluid p_215665_4_, Direction p_215665_5_)
+    {
+        return p_215665_5_ == Direction.DOWN && !p_215665_4_.isIn(FluidTags.WATER);
```
Don't have my compiler open at the moment, so don't remember what this method is for overall given the naming, but is checking against water the correct thing to be doing here given its a custom fluid.

gigaherz replied:
TBH I meant to look at what this is used for, but I forgot, will check.

gigaherz replied:
Did my best. Settled on using isEquivalentTo as default. Mods can override if needed.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java:
```diff
@@ -0,0 +1,243 @@
+
+    @Override
+    public boolean isEquivalentTo(Fluid fluidIn) {
+        return fluidIn == still.get() || fluidIn == flowing.get();
```
Should this also compare any tags that this fluid is in and the given fluid is in? Or is that something that should be in an overwritten method.

gigaherz replied:
Although it may be interesting for a fluid to be able to say eg. that it's "just water" and it should be considered water by any neighbouring water blocks, such as for the level calculations, I don't think it would be correct to just compare tags, since there can be tags that mean something completely different, and we can't tell the difference in code.
I say if someone wants a different test than default, they can override isEquivalentTo.

gigaherz committed:
Use a more generic implementation of func_215665_a. Fluids may still need to override this.

gigaherz committed:
Oops missed some code, and it was important code not to miss, because it uncovered a big derp.

Alpvax reviewed (commented):
JavaDoc for `FluidAttributes#getTranslationKey(FluidStack)` is incorrect, referring to "getUnlocalizedName" instead of "getTranslationKey".

Do we want to add some of the functions from (jabelar's) 1.12 PR to make fluids more/less water-like? i.e. drowning, watering plants, spawning aquatic creatures etc.
Or should we leave that for a different PR?

gigaherz commented:
Extra features are out of the scope of this PR, they can be added on later, the RB is already late enough.
The goal here is to have a solid base that won't require breaking changes.

As for the javadoc, thanks, fixed.

gigaherz committed:
Fix javadoc. Add missing license header.
